### PR TITLE
Introduce C parser library

### PR DESCRIPTION
**Pull request by akyrtzi**
Add a shared library with a C API that provides access to the syntactic parser with callbacks for the inference of raw syntax nodes.
This is primarily intended to be used by SwiftSyntax to speed-up source code parsing for it.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci test OS X platform

File: tools/libSwiftSyntaxParser/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,23 @@
+set(LLVM_EXPORTED_SYMBOL_FILE ${CMAKE_CURRENT_SOURCE_DIR}/libSwiftSyntaxParser.exports)
+
+add_swift_host_library(libSwiftSyntaxParser SHARED
```
Comment (harlanhaskins) [inline tools/libSwiftSyntaxParser/CMakeLists.txt]: Nit: Formatting in this file

Comment (akyrtzi) [reply tools/libSwiftSyntaxParser/CMakeLists.txt]: Fixed.

File: tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  swiftparse_client_node_t parse(const char *source);
+};
+
+class CLibParseActions : public SyntaxParseActions {
+  SourceManager &SM;
+  unsigned BufferID;
+  swiftparse_node_handler_t NodeHandler;
+  swiftparse_node_lookup_t NodeLookup;
```
Comment (nkcsgexi) [inline tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp]: Could we have include them with a `SynParser` here?

File: tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+public:
+  explicit CLibParseActions(SourceManager &sm, unsigned bufID,
+                            swiftparse_node_handler_t hdl,
+                            swiftparse_node_lookup_t lookup)
```
Comment (nkcsgexi) [inline tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp]: Similarly, this constructor can take a `SynParser&` as the third parameter.

File: tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp (hunk 2 above)
Comment (akyrtzi) [reply tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp]: Done.

File: tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp (hunk 3 above)
Comment (akyrtzi) [reply tools/libSwiftSyntaxParser/libSwiftSyntaxParser.cpp]: Done.

**Review by nkcsgexi — APPROVED** [review]

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci test OS X platform

**Pull request merged**